src/main/java/org/sasanlabs/service/vulnerability/ssrf/SSRFVulnerability.java

```diff
@@ -49,4 +61,87 @@
                         false),
                 HttpStatus.BAD_REQUEST);
     }
+
+    @AttackVector(
```

Comment:
We didn't know that the actual implementation of SSRF was wrong, so we added our levels based at the present level. If the re-implementation doesn't need a long time to be done, we can discuss for changing it. Our contribution is limited in time. Can you give us some usecases that you think to implement?

Reply:
yeah, this was added by mistake and hence we are stuck with the wrong SSRF implementation. Thanks for the interest in implementing it.

Sure, I am thinking of a use case where our application will call some endpoint which will list all the projects by SasanLabs but that URL will be controlled from the user interface(UI will only have a button which when pressed shows the projects by SasanLabs (gist page may be) but internally that button will invoke an SSRF endpoint which internally will invoke the provided url in query param)

There can be easily 4-5 levels, 
1. No restriction
2. user is disallowed from file:// extension
2. user is disallowed from localhost
3. user is disallowed from any variant of localhost
4. user is disallowed from metadataservice (you can ignore this one)
5. secure level where whitelist is made
6. secure level where user doesn't control the url.

let me know if have other idea's, as this is the only thing coming to my mind right now.

Reply:
I don't think I really understood the UI interface with a button to press in order to show the projects by SasanLabs. I think that a user needs to enter an URL and this URL needs to be verified in different levels. So, I don't really understand the problem with the current level of SSRF, as we need to enter an URL and when we enter an URL like file://etc/passwd, we have the right information. 

For levels, the two levels that we added can:
- level 2 -> block localhost and 127.0.0.1
- level 3 -> block all IPs present in a black list

Reply:
ok so the current level is not right because of the following issues:
1. It fetches the image and then sends it to UI as a base64 encoded string using data URI which is currently only working with PNG, not with JPEG images or other types.
2. The UI code has some odd check that data should start with "i" which seems unmaintainable in the longer run and doesn't seems like a real world application works this way, atleast i have not seen any application like that.

Also blacklist is not a secured way for handling such usecases, it is very complex to implement the secured level for the current implementation.

Reply:
My suggestion was:
Have a simple button in the UI which loads the list of Projects by SasanLabs or on the page load (html) just go and fetch the projects by SasanLabs but pass the url from user interface. the backend level will take that url and calls it and returns the response from that url.

For level1 to level6 (explained above) will work on taking user input from UI where as level7 will not take the input at all and works directly. However, level6 and level7 both are secure. this way we can test the scanners with false-positive usecases as well.

Reply:
Ok, I will start to re-implement SSRF by taking into consideration your suggestions. We'll keep you posted about our progress.

Reply:
We use a file for testing that we placed in /VulnerableApp. I can push this file if you want to see the results of the tests.

Reply:
My suggestion in the UI side:

we will just have a simple html page showing list of projects by SasanLabs (let's create a github gist for that) and that html page for each of the levels will send a call to backend and backend will call that gist page.
The 2nd last level will validate the url and last one will not take url as input. (both of them are secure)

Why i am not in favour of this design is, because the final goal is not to have a file viewer as you added the whitelist logic.